### ISSUE TITLE
Make Alice granting logic more resilient to potential TX failures/delays

### ISSUE THREAD

**Issue by cygnusv** [nucypher/nucypher#2071]
Due to many factors, a TX may be initially reported as failed, but still be mined eventually. Such factors include gas prices, timeouts, etc.

Currently, our code deals with TX failures, but I'm not sure if it deals with TX that were initially failed, leading to a failure in the parent call, but end up succeeding in the background. 
The following is a hypothetical example: Imagine that Alice uses `grant()`, creates a policy, sends the TX to `PolicyManager`, and after waiting for some time, we timeout and the call to `grant()` fails. It's not impossible, however, that the TX does succeed eventually. Is our `grant()` logic resilient to that? Would it be possible for Alice to recover the situation and resume the operation now that the TX exists? Or is she condemned to live with a zombie policy?

The case of Alice may be the most flagrant, but there are other places to where this can happen. 

Related to #2070

**Comment by KPrasch** [comment]
@cygnusv - This issue may be too broad for closure - the original motive here was mitigation of worker auto-commitments which has been mitigated by new strategies and max price settings, so I'm removing oily rags.

**Comment by cygnusv** [comment]
I'm trying to re-scope this issue to be more Alice-specific and I'm seeing this has ramifications. 

Essentially, our client assumes that the whole granting logic is going to work successfully from start to end, but this is way too optimistic, and also very brittle. There are several places where there may be, at least, some delay, for example in the on-chain TX to set the policy (due to raising gas prices), but in the worst-case there may be no confirmation at all in a reasonable time. In such cases, our logic is bound to fail (which can be acceptable to a certain extent), but what's more worrying is that it's currently not recoverable: e.g., if the policy creation TX clears several hours later, it's obvious that the policy should be valid in the network, which implies (1) having the nodes already storing the KFrag, and (2) have some TMap distribution mechanism ready (being that a deterministic TMap, or a locally-stored TMap, or a TMap already flooded to the network).

For background on this, see the high-level description from @Bogdan here: https://discord.com/channels/411401661714792449/786726745892978719/796922401971241011
```
The thing is, granting is a three-step process: 
1) propose to Ursulas and select n of the ones that agreed
2) post a transaction with the addresses of these n Ursulas
3) send kfrags to Ursulas, along with the transaction hash
So you won't be able to start 3) until your transaction is posted.
Ursulas want to see the transaction to know they are being paid
```

Some ideas/comments:
- We need a clear and fine-grained description of the granting steps. This is in line with concerns about lack of specification of the protocol (thanks @Bogdan for pointing this out in #2418)
- Swapping steps 2 and 3 (as cunningly pointed out by @Noropuck) can make the granting process much more robust to TX failures/delays, at the expense of making the Ursula side more complex. The trade-off can be beneficial IMO, but we still need to consider some potential side-effects, like e.g., opening a DoS vector through KFrag storage: @jMyles (NU Flute Guy) imagine a copy of Nicholas Cage's 2006 remake of "The Wickerman" split as KFrags ("_Not the bees!_" fits inside the space of single KFrag). For the readers, see #2132 and #2301.
- As a counter-argument, I think steps 2 and 3 were swapped in the past, and this was changed a year ago by https://github.com/nucypher/nucypher/pull/1695. Oh boy, it seems it was my suggestion (https://github.com/nucypher/nucypher/issues/121#issuecomment-586514206). @jMyles (NU Flute Guy) I think we need to revisit some of the decisions in #1695.
- In the codebase we use the wording `enact_policy` for sending KFrags to Ursulas (in the `RestMiddleware` class), but I was under the impression that the policy enactment was actually the creation of the policy on-chain. It seems that currently, the latter is considered a sub-step of the latter (see `BlockchainPolicy.enact()`), but that doesn't seem right to me, I think both should be at the same level.

**Comment by vepkenez** [comment]
I'm still a little fuzzy on some of the mechanics but what about the following flow:

1) propose to Ursulas and select n of the ones that agreed
2) receive locally stored treasure map _con kfrags_
     at this point, the alice could just walk away and everything would work... nothing more is needed and she could provide this to Bob and Bob could grant right?
3) (optional) broadcast the transaction and upon the transaction being confirmed,
4) (assuming 3 occurred) _upload_ the treasure map back to any of the ursulas for distribution across the network, they could then check to see if the policy was paid for.  This is like upgrading to the "premium" plan.

If I was trying to get lots of people to use NuCypher, this seems like it would be the simplest MVP (P being "protocol").
Is this workable?

**Comment by cygnusv** [comment]
Excerpt from the discussion on https://discord.com/channels/411401661714792449/786726745892978719/799289823319162890:
> @cygnusv: Tmaps con Kfrags does change the game, thanks for bringing that up!
I'd say that in your idea we need 1, 2 and 3 (so that's not optional)
Step 4 is not needed it you have tmap con kfrags
@vepkenez: So tmaps con kfrags removes the temporal issue and makes it more like “if you want to make bob’s life easier, upload these”
@cygnusv: TMaps con KFrags (we need a good acronym or metaphor for that) removes several problems. If we swap steps 2 and 3 in current logic, there's a potential DoS vector due to KFrag storage, but if there's no need for that, then the steps at grant time are:
>1. propose to Ursulas and select n of the ones that agreed
> 2. post a transaction with the addresses of these n Ursulas
> At retrieve time:
> 3. Bob traverses the list of Ursulas in the TMap
> 4. Each Ursula makes sure the policy is valid (i.e., paid for, not expired, not revoked), and if so, re-encrypts
> 5. Bob decrypts
> Granting is actually simpler

> @vepkenez: Yeah.  This seems like a more viable flow for V1 of our network.  All the complicated tmap storage stuff seems like the kind of thing we might add later
And what are the negatives?  Some direct communication or a side channel need to exist between Alice and bob?
> @cygnusv: Yes, that's the downside: it makes TMap distribution a responsibility of someone else, and in this case, it would seem that only Alice is in position to do it. Since data in the TMap is encrypted, confidentiality is not a problem, but it does have some availability difficulties (although, to be fair, we already are aware that moving that responsibillity to the network comes with a lot of problems as well, e.g., DoS, redundancy, etc). 
> I like to think nucypher more as an application library, so Alice can delegate TMap distribution to the application layer, which most likely, will be shared by Alice and Bob. Essentially, the application layer should implement this side channel. This assumption alleviates many problems. Reasoning about NuCypher as a direct end-user tool overcomplicates the requirements for the network and the nucypher client.

TL;DR: Let's explore TMaps con KFrags (#2234) as a potential improvement of this issue.

**Comment by cygnusv** [comment]
TreasureMaps con KFrags (#2687) fixed this.
